File: go/creational/singleton/connection-pool.go
package singleton

import (
	"fmt"
	"sync"
)

var lock = &sync.Mutex{}

type connectionPool struct {
}

var connPoolInstance *connectionPool

func GetInstance() *connectionPool {
	if connPoolInstance == nil {
		lock.Lock()
		defer lock.Unlock()
		if connPoolInstance == nil {
			fmt.Println("Creating Connection Pool Instance Now")
			connPoolInstance = &connectionPool{}
		} else {
			fmt.Println("Connection Pool Instance already created-1")
		}
	} else {
		fmt.Println("Connection Pool Instance already created-2")
	}
	return connPoolInstance
}

var once sync.Once

func GetInstanceUsingSync() *connectionPool {
	if connPoolInstance == nil {
		once.Do(
			func() {
				fmt.Println("Creating Connection Pool Instance Now")
				connPoolInstance = &connectionPool{}
			})
	} else {
		fmt.Println("Connection Pool Instance already created-2")
	}
	return connPoolInstance
}

File: python/creational/singleton/connectionPool.py
class ConnectionPool(object):
    _instance = None 

    def __new__(cls):
        if cls._instance is None:
            print("Creating the object")
            cls._instance = super(ConnectionPool, cls).__new__(cls)
        return cls._instance

pool1 = ConnectionPool()
print(pool1)

pool2 = ConnectionPool()
print(pool2)
print("Are they the same object?", pool1 is pool2)